clients/src/main/java/org/apache/kafka/common/protocol/types/Schema.java

```diff
@@ -71,7 +71,7 @@ public Struct read(ByteBuffer buffer) {
         Object[] objects = new Object[fields.length];
         for (int i = 0; i < fields.length; i++) {
             try {
-                objects[i] = fields[i].def.type.read(buffer);
+                objects[i] = buffer.position() < buffer.limit() ? fields[i].def.type.read(buffer) : null;
```

Comment:
Did you consider adding `buffer.canRead()` to the for loop criteria?

Reply:
Ah! `hasRemaining`, of course 🤦‍♂️ 
I guess it read more easily if I was explicit about leaving this `null` in the field. But that's a good suggestion.